amoro-ams/amoro-ams-server/src/main/java/org/apache/amoro/server/optimizing/TaskRuntime.java

```diff
@@ -105,6 +105,7 @@ public void complete(OptimizerThread thread, OptimizingTaskResult result) {
           token = null;
           threadId = -1;
         });
+    owner.releaseResourcesIfNecessary();
```

Comment:
How does it avoid the deadlock? The deadlock would happen when 2 thread blocks on `owner.acceptResult(this)` before this line.

Reply:
@rfyu It sounds like what #2684 did.

Reply:
> I think two threads are blocked inside `cancelTasks` within `owner.acceptResult(this)`, so wrapping `cancelTask` within `releaseResourcesIfNecessary` and placing it outside `owner.acceptResult(this)` should indeed solve the deadlock issue

Sorry for late replying.

Just like @rfyu said, place canceling All Tasks out of task and process lock could prevent all lock conflict cases involving canceling all tasks.

When two threads steping into owner.acceptResult(), one will wait for lock anyway, I don't quite get your question

Reply:
I get the idea now. The cancellation is now lock-free. No task lock is needed, thus no deadlock.